src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -107,6 +108,21 @@ def _build_general_name(backend, gn):
         return x509.DirectoryName(
             _build_x509_name(backend, gn.d.directoryName)
         )
+    elif gn.type == backend._lib.GEN_EMAIL:
+        data = backend._ffi.buffer(
+            gn.d.rfc822Name.data, gn.d.rfc822Name.length
+        )[:].decode("ascii")
+        name, address = parseaddr(data)
+        parts = address.split("@")
+        if name or len(parts) > 2:
+            raise ValueError("Invalid rfc822name value")
+        else:
+            if len(parts) == 1:
+                return x509.RFC822Name(address)
+            else:
+                return x509.RFC822Name(
+                    parts[0] + u"@" + idna.decode(parts[1])
+                )
```

Comment:
This would be better represented as an elif chain, also no handling of the 0 case, and I'm not sure I understand the 1 case, more comments please :-)